src/desktop/tabletinput.cpp

```diff
@@ -18,73 +19,223 @@ namespace tabletinput {
+	static constexpr Mode kisTabletWintab = Mode::KIS_TABLET_WINTAB;
+	static constexpr Mode kisTabletWintabRelativePenHack =
+		Mode::KIS_TABLET_WINTAB_RELATIVE_PEN_HACK;
+#	else
+	static Mode kisTabletWinink = qt6Winink;
+	static Mode kisTabletWintab = qt6Wintab;
+	static Mode kisTabletWintabRelativePenHack = qt6Wintab;
+#	endif
+
+	switch(getModeFromSettings(cfg)) {
+	case int(Mode::KIS_TABLET_WININK):
+		return kisTabletWinink;
+	case int(Mode::KIS_TABLET_WINTAB):
+		return kisTabletWintab;
+	case int(Mode::KIS_TABLET_WINTAB_RELATIVE_PEN_HACK):
+		return kisTabletWintabRelativePenHack;
+	case int(Mode::QT5):
+		return qt5;
+	case int(Mode::QT6_WININK):
+		return qt6Winink;
+	case int(Mode::QT6_WINTAB):
+		return qt6Wintab;
+	default:
+		return kisTabletWinink;
+	}
+}
+
+#	ifdef HAVE_KIS_TABLET
+static void resetKisTablet(QApplication *app)
+{
+	if(spontaneousTabletEventFilter) {
+		qApp->removeEventFilter(spontaneousTabletEventFilter);
+		delete spontaneousTabletEventFilter;
+		spontaneousTabletEventFilter = nullptr;
+	}
+	if(kisTabletSupportWin8) {
+		app->removeNativeEventFilter(kisTabletSupportWin8);
+		delete kisTabletSupportWin8;
+		kisTabletSupportWin8 = nullptr;
+	}
+	KisTabletSupportWin::quit();
+}
+
+static void installSpontaneousTabletEventFilter(QApplication *app)
+{
+	spontaneousTabletEventFilter = new SpontaneousTabletEventFilter{app};
+	app->installEventFilter(spontaneousTabletEventFilter);
+}
+
+static void enableKisTabletWinink(QApplication *app)
+{
+	kisTabletSupportWin8 = new KisTabletSupportWin8();
+	if(kisTabletSupportWin8->init()) {
+		installSpontaneousTabletEventFilter(app);
+		app->installNativeEventFilter(kisTabletSupportWin8);
+		inputMode = "KisTablet Windows Ink input";
+		currentMode = Mode::KIS_TABLET_WININK;
+	} else {
+		qWarning("Failed to initialize KisTablet Windows Ink input");
+		delete kisTabletSupportWin8;
+	}
+}
+
+static void enableKisTabletWintab(QApplication *app, bool relativePenModeHack)
+{
+	if(KisTabletSupportWin::init()) {
+		installSpontaneousTabletEventFilter(app);
+		KisTabletSupportWin::enableRelativePenModeHack(relativePenModeHack);
+		if(relativePenModeHack) {
 			inputMode = "KisTablet Wintab input with relative pen mode hack";
 		} else {
-			KisTabletSupportWin::enableRelativePenModeHack(false);
 			inputMode = "KisTablet Wintab input";
 		}
+	} else {
+		qWarning("Failed to initialize KisTablet Wintab input");
 	}
-#elif defined(Q_OS_WIN) && QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
-	Q_UNUSED(app);
-	update(cfg);
-#else
-	// Nothing to do on other platforms.
-	Q_UNUSED(app);
-	Q_UNUSED(cfg);
-#endif
 }
+#	endif
 
-void update(const QSettings &cfg)
+#	if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
+static void enableQt6TabletInput(bool wintab)
 {
-#if defined(Q_OS_WIN) && QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
 	// See by qtbase tests/manual/qtabletevent/regular_widgets/main.cpp
 	using QWindowsApplication = QNativeInterface::Private::QWindowsApplication;
```

Comment:
I think this should be using `#include <QWindowsApplication>` and calling `QGuiApplication::nativeInterface`, they just failed to finish documentation to access the native interface: https://bugreports.qt.io/browse/QTBUG-100077

Reply:
Trying to get this working, it seems like `QWindowsApplication` doesn't work that way, it's a platform interface or something not a native interface. The header doesn't exist anyway and unlike `QNativeInterface::QX11Application`, there appears to be no `QNativeInterface::QWindowsApplication`.

Reply:
Argh, sorry. `QNativeInterface::Private::QWindowsApplication` is still the only one for now. It does look like it should be accessible from `QGuiApplication::nativeInterface` at least, if [this code](https://github.com/wangwenx190/framelesshelper/blob/3b8571401fab94e9eb7bd3dce3b924e08a2784cf/src/core/utils_win.cpp#L1623) and [this code](https://github.com/rgriebl/brickstore/blob/b509923c92a7e3e02a3c8ec0177530ef7665d900/src/common/systeminfo.cpp#L129) are believable, if it matters at all to get rid of one private API.

Reply:
Would definitely be nicer. Changed it to that effect and will give it another try to see if it still works that way.